src/Chapter3.hs

```diff
@@ -1006,6 +1150,59 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+class Actions a where
+  actions :: a -> [Action']
+
+instance Actions (Fighter h d) where
+  actions (Knight' _ _ kActions) = kActions
+  actions (Monster' _ mActions) = mActions
+
+class (Actions a) => Battle a where
+  battle :: a -> a -> [(Action', Action')]
+
+instance Battle (Fighter h d) where
+  battle fighter1 fighter2 = zipDef (Attack' 0) (actions fighter1) (actions fighter2)
+    where
+      zipDef b (x : xs) (y : ys) = (x, y) : zipDef b xs ys
+      zipDef b [] ys = zip (repeat b) ys
+      zipDef b xs [] = zip xs (repeat b)
+
+class Health' a where
+  health :: a -> Int
+
+instance Health' (Fighter a b) where
+  health (Monster' h _) = h
+  health (Knight' h _ _) = h
+
+data Fighter h d
+  = Knight' Int Int [Action']
+  | Monster' Int [Action']
+
+data Action'
+  = Attack' Int
+  | Drink'
+  | CastSpell
+  | RunAway
```

Comment:
Ideally, it's better to have to separate types `MonsterAction` and `KnightAction`, so it won't be possible to accidentally give actions to fighters that can't do them